frontend/src/app/modules/common/dynamic-forms/services/dynamic-form/dynamic-form.service.ts

```diff
@@ -0,0 +1,116 @@
+import { HttpClient, HttpErrorResponse } from "@angular/common/http";
+import { Injectable } from "@angular/core";
+import { FormGroup } from "@angular/forms";
+import { FormlyForm } from "@ngx-formly/core";
+import { Observable } from "rxjs";
+import {
+  catchError,
+  map,
+} from "rxjs/operators";
+import {
+  IOPDynamicForm,
+  IOPForm,
+  IOPFormModel,
+  IFormError,
+} from "../../typings";
+import { DynamicFieldsService } from "core-app/modules/common/dynamic-forms/services/dynamic-fields/dynamic-fields.service";
+@Injectable()
+export class DynamicFormService {
+  dynamicForm:FormlyForm;
+
+  constructor(
+    private _httpClient:HttpClient,
+    private _dynamicFieldsService:DynamicFieldsService,
+  ) {}
+
+  registerForm(dynamicForm:FormlyForm) {
+    this.dynamicForm = dynamicForm;
+  }
+
+  getForm$(url:string): Observable<IOPDynamicForm>{
+    return this._httpClient
+      .post<IOPForm>(
+        url,
+        {},
+        {
+          withCredentials: true,
+          responseType: 'json'
+        }
+      )
+      .pipe(
+        map((formConfig => this._getDynamicFormConfig(formConfig))),
+      );
+  }
+
+  // TODO: Submit to _links.commit?? (pending)
+  submitForm$(formModel:IOPFormModel, resourceEndpoint:string, resourceId?:string) {
+    const modelToSubmit = this._formatModelToSubmit(formModel);
+    const httpMethod = resourceId ? 'patch' : 'post';
+    const url = resourceId ? `${resourceEndpoint}/${resourceId}` : resourceEndpoint;
+
+    return this._httpClient
+      [httpMethod](
+        url,
+        modelToSubmit,
+        {
+          withCredentials: true,
+          responseType: 'json'
+        }
+      )
+      .pipe(
+        catchError((error:HttpErrorResponse) => {
+          this._handleFormErrors(error, this.dynamicForm.form as FormGroup);
+
+          throw error;
+        })
+      );
+  }
+
+  private _getDynamicFormConfig(formConfig:IOPForm):IOPDynamicForm {
+    const formSchema = formConfig._embedded?.schema;
+    const formPayload = formConfig._embedded?.payload;
+    const dynamicForm = {
+      fields: this._dynamicFieldsService.getConfig(formSchema, formPayload),
+      model: this._dynamicFieldsService.getModel(formSchema, formPayload),
+      form: new FormGroup({}),
+    };
+
+    return dynamicForm;
+  }
+
+  private _formatModelToSubmit(formModel:IOPFormModel) {
```

Comment:
Can we set this function and the one below to public? I feel for usecases where we are not using the cookie-cutter dynamic-form approach, we'd need to handle `getForm$` and `submitForm$` manually, but it'd still be good to use the same formatting and error handling approaches, even though we might be only using a subset of the fields and submitting the form in a different way.

Reply:
I'm not sure about this approach because:

- `getForm$`returns the config for a dynamic form and, in order to render a dynamic form, you'd need to repeat all the logic that is in the template of the DynamicFormComponent (create the form, the formlyForm, fetching the config...) which is a lot of work and duplication and it seems to me that leads to inconsistencies:

```
<form [formGroup]="dynamicForm?.form" (submit)="submitForm(dynamicForm?.form.value)">
    <formly-form [form]="dynamicForm?.form"
                 [model]="dynamicForm?.model"
                 [fields]="dynamicForm?.fields">
      <ng-template formlyTemplate let-field>
        <op-form-field *ngFor="let field of dynamicForm?.fields"
                       [label]="field.templateOptions?.label"
                       [inlineLabel]="field.templateOptions?.inlineLabel"
                       [required]="field.templateOptions?.required">
                       
          <formly-field [field]="field" slot=input></formly-field>
          
          <formly-validation-message [field]="field" slot=errors></formly-validation-message>
        </op-form-field>
      </ng-template>
    </formly-form>
    
    <button type="submit"
            class="button -highlight"
            [disabled]="inFlight">
      {{text.save}}
    </button>
  </form>
  ```
  
I see this only needed in order to customize the layout of the form, which could probably be solved with content projection.
  
 - submitForm$: customizing the submit could be done by creating a submitted @Output which outputs the formatted form model. If there are not listeners to the @Output, it would perform the default submitting and error handling.
 
 What do you think?